### PR TITLE
cosmic-{comp,panel,applets}: init

### PR DESCRIPTION
**Pull request by alyssais**
## Description of changes

This is the bare minimum that was necessary for me to have a usable COSMIC environment. (With just cosmic-comp, there's no way to e.g. start an application.)  There are many more components of COSMIC that are not included here.

Since I know a lot of people have been waiting for this, I'll do some expectation management: This is enough to get COSMIC running, but I haven't done any work to integrate COSMIC into NixOS, because running it on NixOS is not my use case.  It will probably work if you add all the packages to environment.systemPackages (because COSMIC relies on some stuff being globally available), and run cosmic-comp and cosmic-panel manually.

Also, COSMIC is pretty usable now, but it's still in very active development.  Please be mindful of that when interacting with upstream.  I've made sure that upstream is okay with us having it packaged in Nixpkgs.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by alyssais** [comment]
> > because running it on NixOS is not my use case
> 
> Out of curiosity: What's your use case then?

https://spectrum-os.org/

**Review by RaitoBezarius — APPROVED** [review]
Built successfully via `nixpkgs-review pr`:

> 4 packages built:
> cosmic-applets cosmic-comp cosmic-comp.debug cosmic-panel

Code looks excellent to me.
Tested multiple binaries without success because I am running X11/i3, so it obviously fails, but binaries do not crash for wrong reasons.

**Comment by alyssais** [comment]
> Tested multiple binaries without success because I am running X11/i3, so it obviously fails, but binaries do not crash for wrong reasons.

It has an X11 backend — I think you should be able to run cosmic-comp. What errors?

**Comment by RaitoBezarius** [comment]
> > Tested multiple binaries without success because I am running X11/i3, so it obviously fails, but binaries do not crash for wrong reasons.
> 
> It has an X11 backend — I think you should be able to run cosmic-comp. What errors?
> 
> EDIT: alternatively, you can of course try running cosmic-comp in its own VT.

I do not have DRM on my X11 server I believe.

```
2023-08-25T13:53:34.095630Z ERROR smithay::backend::egl::ffi: [EGL] 0x3002 (BAD_ACCESS) eglInitialize: EGL_BAD_ACCESS error: In eglInitialize: EGLDisplay (0x55ba11929660): Backend failed to allocate context

2023-08-25T13:53:34.095657Z  WARN smithay::backend::x11: Failed to init X11 surface via egl, falling back to dri3: Failed to initialize EGL. Err: EGL cannot access a requested resource (for example a context is bound in another thread).
2023-08-25T13:53:34.095686Z ERROR smithay::backend::x11: error=The X server is not capable of direct rendering
2023-08-25T13:53:34.095739Z  WARN cosmic_comp::backend: Initializing X11 Backend failed. err=Could not get DRM node used by X server

Caused by:
    The X server is not capable of direct rendering
2023-08-25T13:53:34.096950Z ERROR panic: thread 'main' panicked at 'Failed to initialize any backend! Wayland status: NoCompositorListening X11 status: LibraryOpenError(OpenError { kind: Library, detail: "opening library failed (libXcursor.so.1: cannot open shared object file: No such file or directory); opening library failed (libXcursor.so: cannot open shared object file: No such file or directory)" })': /build/cargo-vendor-dir/winit-0.28.6/src/platform_impl/linux/mod.rs:757
   0: <backtrace::capture::Backtrace as core::default::Default>::default
   1: log_panics::Config::install_panic_hook::{{closure}}
   2: std::panicking::rust_panic_with_hook
   3: std::panicking::begin_panic_handler::{{closure}}
   4: std::sys_common::backtrace::__rust_end_short_backtrace
   5: rust_begin_unwind
   6: core::panicking::panic_fmt
   7: smithay::backend::winit::init
   8: cosmic_comp::backend::winit::init_backend
   9: cosmic_comp::backend::init_backend_auto
  10: cosmic_comp::main
  11: std::sys_common::backtrace::__rust_begin_short_backtrace
  12: main
  13: __libc_start_call_main
  14: __libc_start_main@@GLIBC_2.34
  15: _start
```

**Review by siph — APPROVED** [review]
I was able to get `cosmic-comp` running with wayland.

I couldn't figure out how to run anything from `cosmic` so I tested the `cosmic-applets` with `Hyprland` and everything that didn't depend on anything `cosmic` specific works like expected.

**Comment by amjoseph-nixpkgs** [comment]
> > because running it on NixOS is not my use case

> Out of curiosity: What's your use case then?

I too run a nixpkgs-based, non-NixOS system and am very interested in this package.

**Comment by amjoseph-nixpkgs** [comment]
@ofborg build pkgsCross.aarch64-multiplatform.cosmic-panel

File: pkgs/applications/window-managers/cosmic/panel/default.nix
Comment (amjoseph-nixpkgs) [inline pkgs/applications/window-managers/cosmic/panel/default.nix:30]:
```suggestion
  nativeBuildInputs = [ cargo just pkg-config ];
```

Unfortunately `foo.bar` references (still) don't splice correctly, so you can't use `rustPlatform.cargoSetupHook` in `nativeBuildInputs`.  But you should get the hook automatically from `cargo`.

File: pkgs/applications/window-managers/cosmic/applets/default.nix
Comment (amjoseph-nixpkgs) [inline pkgs/applications/window-managers/cosmic/applets/default.nix:37]:
```suggestion
  nativeBuildInputs = [ cargo just pkg-config ];
```

Unfortunately `foo.bar` references (still) don't splice correctly, so you can't use `rustPlatform.cargoSetupHook` in `nativeBuildInputs`.  But you should get the hook automatically from `cargo`.

File: pkgs/top-level/all-packages.nix
Comment (amjoseph-nixpkgs) [inline pkgs/top-level/all-packages.nix:31062]: Is there a reason why `cosmic-panel` and `cosmic-applets` don't use `buildRustPackage` like `cosmic-comp` does?

There's a lot of important logic in `buildRustPackage` for configuring `cargo` correctly, especially for cross compilation.  Using `stdenv.mkDerivation` means missing out on that, which is why `pkgsCross.aarch64-multiplatform.cosmic-comp` fails to build that way -- cargo isn't even getting the correct `--target` flag.

With the patch below, and my comments above, I can build `pkgsCross.aarch64-multiplatform.cosmic-{comp,applets,panel}`.

```patch
diff --git a/pkgs/applications/window-managers/cosmic/applets/default.nix b/pkgs/applications/window-managers/cosmic/applets/default.nix
index 6c0d2a04461d..3232729c9963 100644
--- a/pkgs/applications/window-managers/cosmic/applets/default.nix
+++ b/pkgs/applications/window-managers/cosmic/applets/default.nix
@@ -4,7 +4,7 @@
 , util-linux
 }:

-stdenv.mkDerivation {
+rustPlatform.buildRustPackage {
   pname = "cosmic-applets";
   version = "unstable-2023-08-23";

@@ -15,7 +15,7 @@ stdenv.mkDerivation {
     hash = "sha256-gf8xf/vzVw3yg+LpmOR+z7VJQtEoMEJ+CFO2vgMmFtM=";
   };

-  cargoDeps = rustPlatform.importCargoLock {
+  cargoLock = {
     lockFile = ./Cargo.lock;
     outputHashes = {
       "accesskit-0.11.0" = "sha256-/6KUCH1CwMHd5YEMOpAdVeAxpjl9JvrzDA4Xnbd1D9k=";
@@ -39,6 +39,7 @@ stdenv.mkDerivation {
   buildInputs = [ dbus glib libxkbcommon pulseaudio wayland ];

   justFlags = [ "--set" "prefix" (placeholder "out") ];
+  dontUseJustBuild = 1;  # we need to let cargo-setup-hook do this; it knows the correct flags

   # Force linking to libwayland-client, which is always dlopen()ed.
   "CARGO_TARGET_${rust.lib.toRustTargetForUseInEnvVars stdenv.hostPlatform}_RUSTFLAGS" = map (a: "-C link-arg=${a}") [
@@ -47,6 +48,12 @@ stdenv.mkDerivation {
     "-Wl,--pop-state"
   ];

+  # upstream justfile `install` target does not understand non-native compilation
+  preInstall = lib.optionalString (rust.lib.toRustTarget stdenv.buildPlatform != rust.lib.toRustTarget stdenv.hostPlatform) ''
+    rm -rf target/release
+    mv target/${rust.lib.toRustTarget stdenv.hostPlatform}/release target/release
+  '';
+
   passthru.updateScript = unstableGitUpdater { };

   meta = with lib; {
diff --git a/pkgs/applications/window-managers/cosmic/panel/default.nix b/pkgs/applications/window-managers/cosmic/panel/default.nix
index a102504ee139..d598c7e717d8 100644
--- a/pkgs/applications/window-managers/cosmic/panel/default.nix
+++ b/pkgs/applications/window-managers/cosmic/panel/default.nix
@@ -3,7 +3,7 @@
 , unstableGitUpdater
 }:

-stdenv.mkDerivation {
+rustPlatform.buildRustPackage {
   pname = "cosmic-panel";
   version = "unstable-2023-08-03";

@@ -14,7 +14,7 @@ stdenv.mkDerivation {
     hash = "sha256-H3QuiP7Og69wm9yCX/uoSG0aQ3B/61q9Sdj+rW4KZMU=";
   };

-  cargoDeps = rustPlatform.importCargoLock {
+  cargoLock = {
     lockFile = ./Cargo.lock;
     outputHashes = {
       "cosmic-config-0.1.0" = "sha256-ENdQkLo4MSH/oe6vFr8qByWehch0nMngBr9/6hF7iVk=";
@@ -31,6 +31,7 @@ stdenv.mkDerivation {
   buildInputs = [ libglvnd libxkbcommon wayland ];

   justFlags = [ "--set" "prefix" (placeholder "out") ];
+  dontUseJustBuild = 1;  # we need to let cargo-setup-hook do this; it knows the correct flags

   # Force linking to libEGL, which is always dlopen()ed.
   "CARGO_TARGET_${rust.lib.toRustTargetForUseInEnvVars stdenv.hostPlatform}_RUSTFLAGS" = map (a: "-C link-arg=${a}") [
@@ -39,6 +40,12 @@ stdenv.mkDerivation {
     "-Wl,--pop-state"
   ];

+  # upstream justfile `install` target does not understand non-native compilation
+  preInstall = lib.optionalString (rust.lib.toRustTarget stdenv.buildPlatform != rust.lib.toRustTarget stdenv.hostPlatform) ''
+    rm -rf target/release
+    mv target/${rust.lib.toRustTarget stdenv.hostPlatform}/release target/release
+  '';
+
   passthru.updateScript = unstableGitUpdater { };

   meta = with lib; {
```

**Review by ghost — APPROVED** [review]
LGTM but the changes below are needed in order for cross compilation to work.  Up to you whether to apply them here, or else leave them to a follow-up PR:

- https://github.com/NixOS/nixpkgs/pull/251555

Only major question/concern: I think using `patchelf` instead of `RUSTFLAGS` will be less of a headache in the long run -- unless there's some reason why we can't use `patchelf --add-needed`?

With the changes below:

Built successfully on:
- [x]  `x86_64-linux`
- [x] `aarch64-linux` (cross from `x86_64-linux`)

These require setting the `use_bindgen` feature flag on `drm-sys`, which `cargo` can't do (but `crate2nix` can!) because it is an **in**direct dependency:  
- [ ] `powerpc64le-linux` 
- [ ] `mips64el-linux` mixed `n32`/`64`  (cross from `x86_64-linux`)

**Comment by alyssais** [comment]
> Only major question/concern: I think using `patchelf` instead of `RUSTFLAGS` will be less of a headache in the long run

Modifying the binary after the fact, rather than just having the linker generate what we want in the first place, feels like way more of a hack/headache to me.

File: pkgs/applications/window-managers/cosmic/applets/default.nix
Comment (alyssais) [reply pkgs/applications/window-managers/cosmic/applets/default.nix:2]: I really wish each applet was it's own package.

File: pkgs/top-level/all-packages.nix
Comment (alyssais) [reply pkgs/top-level/all-packages.nix:31062]: > There's a lot of important logic in `buildRustPackage` for configuring `cargo` correctly, especially for cross compilation. Using `stdenv.mkDerivation` means missing out on that

`buildRustPackage` and other wrappers around `mkDerivation` are horrible and should go away, but you're otherwise correct that not using `cargoBuildHook` means missing out on stuff.  The reason to use just here was that Cargo, not being a proper build system, lacks a concept of installation, and so our options for installing the binary and supporting files are either using `just install`, or duplicating the install logic ourselves.

Is there still important logic you think we'd be missing out on if I fixed the justfile to understand cross targets?

**Comment by amjoseph-nixpkgs** [comment]
> Modifying the binary after the fact, rather than just having the linker generate what we want in the first place, feels like way more of a hack/headache to me.

Well yes, if the build machinery gave us a way to ask for that, sure.  But it really doesn't.

It's your call.

But if we stick with environment variables, we should to use one that applies only to `hostPlatform` linking -- ideally `NIX_CFLAGS_LINK_${suffixSalt}` since that does what you're asking for.  `CARGO_RUSTFLAGS_TARGET_*` with `-Clink` is probably correct too.  I don't think bare `RUSTFLAGS` is.

File: pkgs/top-level/all-packages.nix
Comment (amjoseph-nixpkgs) [reply pkgs/top-level/all-packages.nix:31062]: > `buildRustPackage` and other wrappers around `mkDerivation` are horrible and should go away (because they don't compose and overriding them doesn't work)

I agree, but we need a `rustc-wrapper` in order to do that.

Using `just` for the `installPhase` totally makes sense.

> Is there still important logic you think we'd be missing out on if I fixed the justfile to understand cross targets?

Yeah, the stuff in `cargo-build-hook`:

https://github.com/NixOS/nixpkgs/blob/8c9fc0efefedf6e4a262ccf4d9eba9df7e5ef743/pkgs/build-support/rust/hooks/cargo-build-hook.sh#L31-L45

... which falls into two categories:

1. The environment variables
2. The flags

The other issue is that because we don't have a `NIX_RUSTC_FLAGS_{COMPILE,LINK}`, 
there isn't (currently) a way for you to reference this set of environment variables and flags (https://github.com/NixOS/nixpkgs/pull/247442 fixes this for env vars but not for flags) other than by letting `buildRustPackage` drive the build.

**Comment by amjoseph-nixpkgs** [comment]
Note: `cosmic-comp` builds with `crate2nix`!

If you want to use `nix` to hack on `cosmic-comp`, this is the way to go.  Incremental+distributed builds without "build pollution" strewn all around your `$HOME` directory.

You need the very latest `crate2nix` plus:

- https://github.com/kolloch/crate2nix/pull/288/
- https://github.com/NixOS/nixpkgs/pull/251534

`git clone` upstream's source and run `crate2nix generate` to get `Cargo.nix` and `crate-hashes.json`.  Keep those two files, discard the rest of the upstream source. These two files are what would go into a PR if we allowed it.  Then `nix-build` the file below, passing `--arg 'import <nixpkgs>'` or similar.

```
{ pkgs }:
(import ./Cargo.nix { inherit pkgs; })
.rootCrate
.build
.overrideAttrs {
  buildInputs = with pkgs; [
    libinput
    mesa
    libxkbcommon
  ];
}
```

This just gets it built (without using `cargo`).  For a nice finished package you still want the `installPhase`, `meta`, and other trimmings found in this PR.

The big wins with `crate2nix` are:

- Overrides work!  Deep down in the dependency tree too!
- Incremental builds, at crate granularity.  Great for hacking.  Build in `/nix/store` as fast as incrementally as `cargo` does in `$HOME`.
- Distributed builds: each crate is a derivation, they build in parallel on different machines.  This lets you build *much faster* than `cargo` does if you have more than one builder.

**Comment by amjoseph-nixpkgs** [comment]
Hey I'm about to leave town for a while, so let's just postpone fixing cross to a later PR.  I've marked all but one of my suggested changes as "resolved", and that last one might not matter.  Thanks for writing this!

**Comment by RaitoBezarius** [comment]
Result of `nixpkgs-review pr 251365` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>cosmic-applets</li>
    <li>cosmic-comp</li>
    <li>cosmic-comp.debug</li>
    <li>cosmic-panel</li>
  </ul>
</details>

**Comment by Nyabinary** [comment]
What's the status of this pr?

**Comment by colemickens** [comment]
@amjoseph-nixpkgs can you [email](mailto:cole.mickens@gmail.com), or matrix me (@colemickens:matrix.org)? I am quite intrigued at the Crate2nix and have a little spike related to this PR and your comments that you might be interested in, and that I could use your advice on.

**Comment by Nyabinary** [comment]
https://github.com/NixOS/nixpkgs/pull/247442 has been merged into staging.

**Comment by alyssais** [comment]
I've updated this to take advantage of https://github.com/NixOS/nixpkgs/pull/257411, which was the last change I wanted to make.  I think this is ready now.

**Comment by Nyabinary** [comment]
Should a tracking issue be made for COSMIC in general?

**Comment by alyssais** [comment]
Wouldn't that be https://github.com/NixOS/nixpkgs/issues/199563?

**Comment by Nyabinary** [comment]
@alyssais Whoops just saw your comment now, I made #259641

**Comment by amjoseph-nixpkgs** [comment]
> @amjoseph-nixpkgs can you [email](mailto:cole.mickens@gmail.com), or matrix me (@colemickens:matrix.org)? I am quite intrigued at the Crate2nix and have a little spike related to this PR and your comments that you might be interested in, and that I could use your advice on.

@colemickens I'm digging myself out from under a huge backlog right now.  Give me some time to get caught up.

**Comment by KiaraGrouwstra** [comment]
i'm not sure how different these OSs are, but as expected i didn't quite manage to get this up and running on nixos yet:

- `cosmic-panel`: running separately from the command-line yields `ERROR Failed to connect to the notifications daemon Ok(Err(environment variable not found))`
- `cosmic-applets`: applications listed in `ls /nix/store/*-cosmic-applets*/bin/` seem to do nothing when run from the command-line 
- `cosmic-comp`: running yields `ERROR smithay::xwayland::xserver: XWayland crashed at startup, will not try to restart it.`

i'd be curious if anyone has had more success there. :pray:

**Comment by MtFBella109** [comment]
I tried it, I added all the cosmic-packages to environment.Systempackages, as I try to start cosmic-comp in a tty, I get a few Errors, that it can't read a few Configs, like something with xdg and input and doesn't start at all

**Comment by amjoseph-nixpkgs** [comment]
I'm still waiting for the long-overdue merge of `staging` into `master`; this is one of the many tasks I have which are blocked on that.

**Pull request merged**
